Nodejs/Product/TestAdapter/TestExecutorWorker.cs

```diff
@@ -0,0 +1,390 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.NetworkInformation;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.NodejsTools.TestAdapter.TestFrameworks;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+using Microsoft.VisualStudioTools;
+using Microsoft.VisualStudioTools.Project;
+using Newtonsoft.Json;
+
+namespace Microsoft.NodejsTools.TestAdapter
+{
+    internal sealed class TestExecutorWorker
+    {
+        private static readonly Version Node8Version = new Version(8, 0);
+
+        //get from NodeRemoteDebugPortSupplier::PortSupplierId
+        private static readonly Guid NodejsRemoteDebugPortSupplierUnsecuredId = new Guid("{9E16F805-5EFC-4CE5-8B67-9AE9B643EF80}");
+
+        private readonly ManualResetEvent cancelRequested = new ManualResetEvent(false);
+        private readonly ManualResetEvent testsCompleted = new ManualResetEvent(false);
+
+        private readonly IFrameworkHandle frameworkHandle;
+        private readonly IRunContext runContext;
+        private List<TestCase> currentTests;
+        private ProcessOutput nodeProcess;
+        private TestResult currentResult;
+        private ResultObject currentResultObject;
+
+        public TestExecutorWorker(IRunContext runContext, IFrameworkHandle frameworkHandle)
+        {
+            this.frameworkHandle = frameworkHandle;
+            this.runContext = runContext;
+        }
+
+        public void Cancel()
+        {
+            //let us just kill the node process there, rather do it late, because VS engine process 
+            //could exit right after this call and our node process will be left running.
+            KillNodeProcess();
+            this.cancelRequested.Set();
+        }
+
+        /// <summary>
+        /// This is the equivalent of "RunAll" functionality
+        /// </summary>
+        /// <param name="sources">Refers to the list of test sources passed to the test adapter from the client.  (Client could be VS or command line)</param>
+        /// <param name="runContext">Defines the settings related to the current run</param>
+        /// <param name="frameworkHandle">Handle to framework.  Used for recording results</param>
+        public void RunTests(IEnumerable<string> sources, ITestDiscoverer discoverer)
```

Comment:
Is there some reason to believe a run isn't already in progress?

Reply:
Added calls to cancel